[autograd.Function] Add docs on the functorch interaction

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91452

This PR:
- Updates autograd.Function.forward docs to reflect how you either
  define a forward with ctx or a separate forward and setup_context
- Updates the "Extending Autograd" docs to suggest the usage of
  autograd.Function with separate forward and setup_context. This should
  be the default because there is a low barrier to go from this to
  an autograd.Function that is fully supported by functorch transforms.
- Adds a new "Extending torch.func with autograd.Function" doc that
  explains how to use autograd.Function with torch.func. It also
  explains how to use generate_vmap_rule and how to manually write a
  vmap staticmethod.

While writing this, I noticed that the implementation of
setup_context staticmethod/generate_vmap_rule/vmap staticmethod are a
bit inconsistent with the other method/attributes on autograd.Function:
- https://github.com/pytorch/pytorch/issues/91451
- I'm happy to fix those if we think it is a problem, either in this PR
  or a followup (this PR is getting long, I want some initial docs
  out that I can point early adopters at, and fixing the problems in the
  future isn't really BC-breaking).

Test Plan:
- view docs preview

zou3519 committed:
[autograd.Function] Add docs on the functorch interaction

This PR:
- Updates autograd.Function.forward docs to reflect how you either
  define a forward with ctx or a separate forward and setup_context
- Updates the "Extending Autograd" docs to suggest the usage of
  autograd.Function with separate forward and setup_context. This should
  be the default because there is a low barrier to go from this to
  an autograd.Function that is fully supported by functorch transforms.
- Adds a new "Extending torch.func with autograd.Function" doc that
  explains how to use autograd.Function with torch.func. It also
  explains how to use generate_vmap_rule and how to manually write a
  vmap staticmethod.

While writing this, I noticed that the implementation of
setup_context staticmethod/generate_vmap_rule/vmap staticmethod are a
bit inconsistent with the other method/attributes on autograd.Function:
- https://github.com/pytorch/pytorch/issues/91451
- I'm happy to fix those if we think it is a problem, either in this PR
  or a followup (this PR is getting long, I want some initial docs
  out that I can point early adopters at, and fixing the problems in the
  future isn't really BC-breaking).

Test Plan:
- view docs preview

[ghstack-poisoned]

soulitzer commented on torch/autograd/function.py:
```diff
@@ -296,12 +296,36 @@ def __init__(cls, name, bases, attrs):
 
-        It must accept a context ctx as the first argument, followed by any
-        number of arguments (tensors or other types).
+        - You must also define a setup_context staticmethod to handle setting up the
```
nit: maybe add a point about: "forward no longer takes context ctx as the first argument."

soulitzer commented on docs/source/notes/extending.rst:
```diff
@@ -72,6 +73,12 @@ Take the following steps:
+  not accept ``ctx`` and a ``setup_context`` method where the ``ctx`` modification happens.
+  In general the separate :meth:`~Function.forward` and ``setup_context`` is closer to how
+  PyTorch native operations work and therefore more composable with various PyTorch subsystems.
+  See :ref:`combining-forward-context` for more details
```
We mention that we shouldn't do compute inside setup_context in an example somewhere above, but maybe we should mention that here too.

soulitzer commented on docs/source/notes/extending.func.rst:
```diff
@@ -0,0 +1,479 @@
+  have it work with function transforms. That is, the :class:`torch.autograd.Function`'s
+  forward/backward/etc calls into functions from other systems like C++, CUDA, numpy.
+- you wish to specify custom gradient rules, like
+  JAX's `custom_vjp/custom_jvp <https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html>`_
```
Do we want to mention that there might be unexpected behavior if the user tries to mix the two use cases. For example if user specifies a custom vmap rule AND wants to do custom gradients, the former can override the latter.

zou3519 replied:
Is that unexpected behavior though?

If I'm a user, and I write a forward() + backward() using only PyTorch operations, and I want custom vmap behavior, then I would write a custom vmap rule rather than use generate_vmap_rule.

Though I agree we should mention that this case is possible. I'll put it below in the "torch.vmap Support" section.

soulitzer replied:
yeah nah, not unexpected behavior, but probably tricky enough to deserve a mention

zou3519 replied:
Added a mention, as well as a small note about the grad(vmap(Func)) semantics

soulitzer commented on docs/source/notes/extending.rst:
```diff
@@ -233,6 +289,54 @@ See :ref:`grad-check` for more details on finite-difference gradient comparisons
+We recommend the second option (separate :meth:`~Function.forward` and ``setup_context``)
+because that is closer to how PyTorch native operations are implemented and it composes
+with :mod:`torch.func` transforms. However, we plan to support both approaches going forward;
+combining :meth:`~Function.forward` with ``setup_context`` leads to more flexibility.
```
nit: we aren't super specific about why combining is more flexible here (i.e. allow for saving intermediates?)

zou3519 replied:
yeah, saving intermediates without promoting them to be outputs

soulitzer commented on docs/source/notes/extending.func.rst:
```diff
@@ -0,0 +1,479 @@
+Here's how to define the ``vmap`` staticmethod:
+
+- the signature is ``vmap(info, in_dims: Tuple[Optional[int]], *args)``, where
+  ``*args`` is the same as the args to :meth:`~Function.forward`.
```
should vmap take `**kwargs` for consistency since forward does (though they have to be passed positionally because apply doesn't take true kwargs)

zou3519 replied:
I might take this as a follow-up item. I am not sure:
- `setup_context(ctx, inputs, output)` takes in `inputs` that were directly passed to forward
- If there were ANY unspecified values that are default kwargs, then we have a problem -- they're not passed to setup_context, so setup_context doesn't get their default value (if it needs to save them for backward)
- So the conclusion is that autograd.Function with setup_context has some really awkward handling for default kwargs.

I'm inclined to say in the docs that autograd.Function with setup_context doesn't support `**kwargs`

zou3519 replied:
I end up changing the documentation to suggest that default args and keywords args be passed via wrapper so it's clearer how to handle them, but not changing the suggested signature of the vmap staticmethod

soulitzer commented on docs/source/notes/extending.func.rst:
```diff
@@ -0,0 +1,479 @@
+  It is ``None`` if the arg is not a Tensor or if the arg is not being vmapped over,
+  otherwise, it is an integer specifying what dimension of the Tensor is being vmapped
+  over.
+- ``info`` is a collection of additional metadata that may be helpful. For example,
```
Do we want to mention all the metadata here? (or link to somewhere where one could find that information)

zou3519 replied:
Sure

soulitzer reviewed (approved):
Nice!

soulitzer commented:
Will take another look once the docs render

zou3519 commented:
> Will take another look once the docs render

Thanks!

zou3519 commented on docs/source/notes/extending.rst:
```diff
@@ -205,6 +226,41 @@ And here, we optimize the above example by calling set_materialize_grads(False):
             # Gradients of non-Tensor arguments to forward must be None.
             return grad_output * ctx.constant, None
 
+If you need any "intermediate" Tensors computed in :meth:`~Function.forward` to be saved,
```
Sgtm

zou3519 committed:
Update on "[autograd.Function] Add docs on the functorch interaction"


This PR:
- Updates autograd.Function.forward docs to reflect how you either
  define a forward with ctx or a separate forward and setup_context
- Updates the "Extending Autograd" docs to suggest the usage of
  autograd.Function with separate forward and setup_context. This should
  be the default because there is a low barrier to go from this to
  an autograd.Function that is fully supported by functorch transforms.
- Adds a new "Extending torch.func with autograd.Function" doc that
  explains how to use autograd.Function with torch.func. It also
  explains how to use generate_vmap_rule and how to manually write a
  vmap staticmethod.

While writing this, I noticed that the implementation of
setup_context staticmethod/generate_vmap_rule/vmap staticmethod are a
bit inconsistent with the other method/attributes on autograd.Function:
- https://github.com/pytorch/pytorch/issues/91451
- I'm happy to fix those if we think it is a problem, either in this PR
  or a followup (this PR is getting long, I want some initial docs
  out that I can point early adopters at, and fixing the problems in the
  future isn't really BC-breaking).

Test Plan:
- view docs preview

[ghstack-poisoned]

zou3519 committed:
Update on "[autograd.Function] Add docs on the functorch interaction"


This PR:
- Updates autograd.Function.forward docs to reflect how you either
  define a forward with ctx or a separate forward and setup_context
- Updates the "Extending Autograd" docs to suggest the usage of
  autograd.Function with separate forward and setup_context. This should
  be the default because there is a low barrier to go from this to
  an autograd.Function that is fully supported by functorch transforms.
- Adds a new "Extending torch.func with autograd.Function" doc that
  explains how to use autograd.Function with torch.func. It also
  explains how to use generate_vmap_rule and how to manually write a
  vmap staticmethod.

While writing this, I noticed that the implementation of
setup_context staticmethod/generate_vmap_rule/vmap staticmethod are a
bit inconsistent with the other method/attributes on autograd.Function:
- https://github.com/pytorch/pytorch/issues/91451
- I'm happy to fix those if we think it is a problem, either in this PR
  or a followup (this PR is getting long, I want some initial docs
  out that I can point early adopters at, and fixing the problems in the
  future isn't really BC-breaking).

Test Plan:
- view docs preview

[ghstack-poisoned]

zou3519 committed:
Update on "[autograd.Function] Add docs on the functorch interaction"


This PR:
- Updates autograd.Function.forward docs to reflect how you either
  define a forward with ctx or a separate forward and setup_context
- Updates the "Extending Autograd" docs to suggest the usage of
  autograd.Function with separate forward and setup_context. This should
  be the default because there is a low barrier to go from this to
  an autograd.Function that is fully supported by functorch transforms.
- Adds a new "Extending torch.func with autograd.Function" doc that
  explains how to use autograd.Function with torch.func. It also
  explains how to use generate_vmap_rule and how to manually write a
  vmap staticmethod.

While writing this, I noticed that the implementation of
setup_context staticmethod/generate_vmap_rule/vmap staticmethod are a
bit inconsistent with the other method/attributes on autograd.Function:
- https://github.com/pytorch/pytorch/issues/91451
- I'm happy to fix those if we think it is a problem, either in this PR
  or a followup (this PR is getting long, I want some initial docs
  out that I can point early adopters at, and fixing the problems in the
  future isn't really BC-breaking).

Test Plan:
- view docs preview

[ghstack-poisoned]

zou3519 commented:
Docs preview is up if folks wanted to take another pass: https://docs-preview.pytorch.org/91452/notes/extending.func.html

zou3519 committed:
Update on "[autograd.Function] Add docs on the functorch interaction"


This PR:
- Updates autograd.Function.forward docs to reflect how you either
  define a forward with ctx or a separate forward and setup_context
- Updates the "Extending Autograd" docs to suggest the usage of
  autograd.Function with separate forward and setup_context. This should
  be the default because there is a low barrier to go from this to
  an autograd.Function that is fully supported by functorch transforms.
- Adds a new "Extending torch.func with autograd.Function" doc that
  explains how to use autograd.Function with torch.func. It also
  explains how to use generate_vmap_rule and how to manually write a
  vmap staticmethod.

While writing this, I noticed that the implementation of
setup_context staticmethod/generate_vmap_rule/vmap staticmethod are a
bit inconsistent with the other method/attributes on autograd.Function:
- https://github.com/pytorch/pytorch/issues/91451
- I'm happy to fix those if we think it is a problem, either in this PR
  or a followup (this PR is getting long, I want some initial docs
  out that I can point early adopters at, and fixing the problems in the
  future isn't really BC-breaking).

Test Plan:
- view docs preview

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "unrelated failure"
